### ISSUE TITLE
Once authorized by Oauth2-Proxy via Gitlab, every 120 seconds I have to sign back.

### ISSUE THREAD

**Issue by rdecuir** [oauth2-proxy/oauth2-proxy#731]
## Issue
I'm using NGINX as a reverse proxy with Oauth2-Proxy hitting a Gitlab instance for authentication.  Every 120 seconds from when I signed in last, I end up having to sign in again.

Workflow:
1.  Hit the proxy url and am prompted to log in using oauth2-proxy via "Sign in with Gitlab" button.
2.  Once I hit the button and enter my credentials I then am redirected to our splash page with a directory of the UIs urls behind the proxy.
3.  After sitting at the splash page for 120 seconds, if I hit refresh I am redirected back to the page that contains the "Sign in with Gitlab" button and I have to get authenticated again.
4.  If I have navigated away from the splash page to one of the UIs we want users to have access to, after 120 seconds I start getting 401 Unauthorized errors and have to manual navigate back to the proxy url which displays the "Sign in with Gitlab" button and I have to get authenticated again.

## Context
Main goal is to place our UIs behind a NGINX reverse proxy, authenticating with Oauth2-proxy using our own instance of Gitlab, and have users log in only once per day giving access to all UIs behind the proxy.

## Expected Behavior
I expect that since I am using the defaults, and the cookie aren't expected to expire for a week, my session should last at least that long after I sign in using Oauth2 and Gitlab.

## Current Behavior
Every 120 seconds, I have to "Sign in with Gitlab" again. If I am on on of the UIs behind the proxy I start getting 401 unauthorized errors and am forced to go back to the proxy url to sign back in.

I have made sure, using the web browser console and network dev tools, that I am getting a "_oauth2_proxy:" cookie, and see that it expires after 168 hours.

## Your Environment
oauth2-proxy v6.0.0

Oauth2 command:
```
/opt/oauth2-proxy/oauth2-proxy \
    --cookie-samesite=none \
    --email-domain="*" \
    --upstream=http://proxy.company.org/ \
    --cookie-secret={redacted} \
    --cookie-secure=true \
    --pass-authorization-header=true \
    --provider=gitlab \
    --client-id={redacted} \
    --client-secret={redacted} \
    --oidc-issuer-url=https://gitlab.company.org \
    --gitlab-group=person2 \
    --set-xauthrequest=true \
    --redirect-url=https://proxy.company.org/oauth2/callback
```

NGINX Proxy configuration:
```
server {
  listen 443 ssl;
  server_name proxy.company.org;

  ssl_certificate /etc/pki/proxy.pem;
  ssl_certificate_key /etc/pki/private.key;

  location /oauth2/ {
    proxy_pass       http://127.0.0.1:4180;
    proxy_set_header Host                    $host;
    proxy_set_header X-Real-IP               $remote_addr;
    proxy_set_header X-Scheme                $scheme;
    proxy_set_header X-Auth-Request-Redirect $request_uri;
    proxy_set_header Content-Length   "";
    proxy_pass_request_body           off;
  }

  location / {
    auth_request /oauth2/auth;
    error_page 401 = /oauth2/sign_in;

    # pass information via X-User and X-Email headers to backend,
    # requires running with --set-xauthrequest flag
    auth_request_set $user   $upstream_http_x_auth_request_user;
    auth_request_set $email  $upstream_http_x_auth_request_email;
    proxy_set_header X-User  $user;
    proxy_set_header X-Email $email;


    # if you enabled --cookie-refresh, this is needed for it to work with auth_request
    auth_request_set $auth_cookie $upstream_http_set_cookie;
    add_header Set-Cookie $auth_cookie;
  
    root /usr/share/nginx/html;
  }

  #-------------UI Endpoints Below

  location = /jupyterhub {
    rewrite ^/(.*)$ $1/ permanent;
  }

  location /jupyterhub {
    auth_request /oauth2/auth;
    try_files $uri @proxy_jupyterhub;
  }

  location @proxy_jupyterhub {
    #rewrite /groot(.*) $1  break;
    proxy_read_timeout 300s;
    proxy_pass https:///jupyter.company.org6:9443;
    # pass some extra stuff to the backend
    proxy_set_header Host $host;
    proxy_set_header X-Real-Ip $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
  }
  location ~ /jupyterhub/(user/[^/]*)/(api/kernels/[^/]+/channels|terminals/websocket)(.*) {
    proxy_pass            https://jupyter.company.org:9443;
    proxy_set_header      Host $host;
    # websocket support
    proxy_http_version    1.1;
    proxy_set_header      Upgrade "websocket";
    proxy_set_header      Connection "Upgrade";
    proxy_read_timeout    86400;
  }

  location /kibana {
    auth_request /oauth2/auth;
    proxy_pass http://search.company.org:5601;
  }

```
Also, I have noticed every time I am signing into gitlab, I am being redirected to a 500 error page at first but I just have to go back to the sign in button page and refresh and hit the button again and am able to go through. This seems like a another issue, but figured I would mention it.

**Comment by JoelSpeed** [comment]
@rdecuir Did you manage to work out your issue? Just wondering if you could post the solution in case anyone else comes across this at a later date

**Comment by rdecuir** [comment]
Issue was using the prebuild binary oauth2-proxy-v6.0.0.linux-arm64.go1.14.2. Once I installed Git/Go and build the oauth2-proxy myself, it worked with my configurations and the sessions now last as long as the cookie is valid for.

I wanted to avoid installing git and go on my minimal VM, so I opted for the prebuild binary originally. Everything seemed to run as expected, up until noticing the above issue. After trying everything online I could find that might have helped and nothing worked, I opted to install the needed packages to build directly from you git repo.

Not sure, but maybe there was a fix or something in the git repo that was not in the prebuilt binary.
